locations/middlewares/track_sources.py

```diff
@@ -0,0 +1,36 @@
+from urllib.parse import urlparse
+
+from scrapy import Spider
+from scrapy.item import Item
+
+
+class TrackSourcesMiddleware:
+    """
+    Track the URLs used for each generated item,
+    and overall stats per hostname.
+    """
+
+    def _process_item(self, response, item: Item, spider: Spider):
```

Comment:
Do we want to assert this is a Feature, which is our extension to an Item?
Otherwise there's a slight risk Item won't have the key "extras".

Found via https://github.com/alltheplaces/alltheplaces/pull/9253#issuecomment-2267395207

Reply:
Other pipelines require extras, at this stage if a spider isn't outputting it it's an issue with the spider. But this might now be the first an item goes through?

On Item vs Feature, Feature is an Item.  All our pipelines require them even though scrapy can do dicts and other things. Minor preference for Item incase we add other types in future but no strong preference.